File: src/main/java/at/yawk/katbot/ForumListener.java
package at.yawk.katbot;

import java.io.IOException;
import java.net.URI;
import java.util.*;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;
import javax.inject.Inject;
import javax.inject.Provider;
import javax.inject.Singleton;
import lombok.RequiredArgsConstructor;
import lombok.Value;
import lombok.extern.slf4j.Slf4j;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.kitteh.irc.client.library.Client;
import org.kitteh.irc.client.library.element.Channel;

/**
 * @author yawkat
 */
@Slf4j
@Singleton
class ForumListener {
    @Inject ScheduledExecutorService executor;
    @Inject Config config;
    // lazy init
    @Inject Provider<UrlShortener> urlShortener;
    @Inject Client client;

    private final Map<URI, ForumHolder> forums = new HashMap<>();

    public void start() {
        executor.scheduleWithFixedDelay(() -> {
            try {
                pollAll();
            } catch (Exception e) {
                log.error("Failed to poll forums", e);
            }
        }, 0, 1, TimeUnit.MINUTES);
    }

    private synchronized void pollAll() throws IOException {
        for (Map.Entry<URI, ForumConfiguration> entry : config.getForums().entrySet()) {
            forums.computeIfAbsent(entry.getKey(), k -> new ForumHolder(k, entry.getValue())).poll();
        }
    }

    static void sendToChannels(Client client, List<String> channels, String message) {
        for (String channelName : channels) {
            Optional<Channel> channelOptional = client.getChannel(channelName);
            if (!channelOptional.isPresent()) {
                log.warn("Could not find channel {}", channelName);
                continue;
            }
            channelOptional.get().sendMessage(message);
        }
    }

    @RequiredArgsConstructor
    private class ForumHolder {
        private final URI uri;
        private final ForumConfiguration configuration;

        private final Set<Integer> sentThreadIds = new HashSet<>();
        private boolean firstPass = true;

        private void poll() throws IOException {
            assert Thread.holdsLock(ForumListener.this);

            List<ThreadInfo> threads = getThreads();
            for (ThreadInfo thread : threads) {
                boolean added = sentThreadIds.add(thread.getId());
                if (thread.replyCount <= 0 && added && !firstPass) {
                    String message = Template.parse(configuration.messagePattern)
                            .set("title", thread.getTitle())
                            .set("author", thread.getAuthor())
                            .set("uri", thread.getUri().toString())
                            .set("uri.short", () -> {
                                try {
                                    return urlShortener.get().shorten(thread.getUri()).toString();
                                } catch (Exception e) {
                                    throw new RuntimeException(e);
                                }
                            })
                            .finish();

                    log.info("Sending forum update '{}' to {} channels", message, configuration.channels.size());

                    sendToChannels(client, configuration.channels, message);
                }
            }
            firstPass = false;
        }

        private List<ThreadInfo> getThreads() throws IOException {
            List<ThreadInfo> threads = new ArrayList<>();
            Document document = Jsoup.connect(uri.toString()).get();
            for (Element element : document.select(".all_threads_container .thread_content")) {
                if (element.hasClass("thread_separator")) { continue; }

                Element titleTag = element.select(".name > a").first();
                String replyCountText = element.select(".replycount").text().replaceAll("[^\\d]", "");
                String href = titleTag.attr("href");
                threads.add(new ThreadInfo(
                        Integer.parseInt(href.substring(href.indexOf('/') + 1, href.indexOf('-'))),
                        URI.create(titleTag.absUrl("href")),
                        titleTag.text(),
                        replyCountText.isEmpty() ? 0 : Integer.parseInt(replyCountText),
                        element.select(".topicstart > a").text()
                ));
            }
            return threads;
        }
    }

    @Value
    private static class ThreadInfo {
        private final int id;
        private final URI uri;
        private final String title;
        private final int replyCount;
        private final String author;
    }

    @Value
    public static class ForumConfiguration {
        private final List<String> channels;
        private final String messagePattern;
    }
}

File: src/main/kotlin/at/yawk/katbot/ForumListener.kt
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 */

package at.yawk.katbot

import org.jsoup.Jsoup
import org.slf4j.LoggerFactory
import java.net.URI
import java.util.*
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Provider

/**
 * @author yawkat
 */
class ForumListener @Inject constructor(
        val executor: ScheduledExecutorService,
        val config: Config,
        // lazy init
        val urlShortener: Provider<UrlShortener>,
        val ircProvider: IrcProvider
) {

    companion object {
        val log = LoggerFactory.getLogger(ForumListener::class.java)
    }

    private val forums = HashMap<URI, ForumHolder>()

    fun start() {
        executor.scheduleWithFixedDelay({
            try {
                pollAll()
            } catch (e: Exception) {
                log.error("Failed to poll forums", e)
            }
        }, 0, 1, TimeUnit.MINUTES)
    }

    @Synchronized
    private fun pollAll() {
        for (entry in config.forums.entries) {
            forums.computeIfAbsent(entry.key) { it -> ForumHolder(it, entry.value) }
                    .poll()
        }
    }

    private inner class ForumHolder(
            val uri: URI,
            val configuration: ForumConfiguration
    ) {

        private val sentThreadReplyCounts = HashMap<Int, Int>()
        private var firstPass = true

        fun poll() {
            assert(Thread.holdsLock(this@ForumListener))

            val threads = fetchThreads()
            for (thread in threads) {
                val oldReplyCount = sentThreadReplyCounts.put(thread.id, thread.replyCount)
                if (thread.replyCount != oldReplyCount && !firstPass) {
                    val message = configuration.messagePattern
                            .with("title", thread.title)
                            .with("author", thread.author)
                            .with("uri", thread.uri.toString())
                            .with("uri.short", { urlShortener.get().shorten(thread.uri).toString() })
                    log.info("Sending forum update '{}' to {} channels", message, configuration.channels.size)
                    message.sendTo(ircProvider.findChannels(configuration.channels))
                }
            }
            firstPass = false
        }

        private fun fetchThreads(): ArrayList<ThreadInfo> {
            val threads = ArrayList<ThreadInfo>()
            val document = Jsoup.connect(uri.toString()).get()
            for (element in document.select(".all_threads_container .thread_content")) {
                if (element.hasClass("thread_separator")) {
                    continue
                }

                val titleTag = element.select(".name > a").first()
                val replyCountText = element.select(".replycount").text().replace("[^\\d]".toRegex(), "")
                val href = titleTag.attr("href")
                threads.add(ThreadInfo(
                        Integer.parseInt(href.substring(href.indexOf('/') + 1, href.indexOf('-'))),
                        URI.create(titleTag.absUrl("href")),
                        titleTag.text(),
                        if (replyCountText.isEmpty()) 0 else Integer.parseInt(replyCountText),
                        element.select(".topicstart > a").text()))
            }
            return threads
        }
    }

    data class ThreadInfo(
            val id: Int,
            val uri: URI,
            val title: String,
            val replyCount: Int,
            val author: String
    )

    data class ForumConfiguration(
            val channels: List<String>,
            val messagePattern: at.yawk.katbot.template.Template
    )
}